code/controllers/subsystems/typing.dm

```diff
@@ -0,0 +1,208 @@
+SUBSYSTEM_DEF(typing)
+	name = "Typing"
+	flags = SS_BACKGROUND | SS_NO_INIT
+	wait = 0.5 SECONDS
+
+	/// The skin control to poll for TYPING_STATE_INPUT status.
+	var/const/INPUT_HANDLE = "outputwindow.input"
+
+	var/const/INFLIGHT_TIMEOUT = 5 SECONDS
+
+	/// The status entry index of the related client's typing indicator visibility preference.
+	var/const/INDEX_PREFERENCE = 1
+
+	/// The status entry index of the inflight state.
+	var/const/INDEX_INFLIGHT = 2
+
+	/// The status entry index of the timeout threshold.
+	var/const/INDEX_TIMEOUT = 3
+
+	/// The status entry index of the input bar typing state.
+	var/const/INDEX_INPUT_STATE = 4
+
+	/// The status entry index of the verb input typing state.
+	var/const/INDEX_VERB_STATE = 5
+
+	/// The highest index in a status entry.
+	var/const/MAX_INDEX = 5
+
+	/*
+	* A list of (ckey = list(
+			preference = 0|1,
+			inflight = 0|1,
+			timeout = num,
+			istyping_input = 0|1,
+			istyping_hotkey = 0|1
+		), ...)
+		See .proc/get_entry for details.
+	*/
+	var/static/list/status = list()
+
+	/// Matches input bar verbs that should set TYPING_STATE_INPUT.
+	var/static/regex/match_verbs = regex("^(Me|Say) +\"?\\w+")
+
+	/// A list of clients waiting to be polled for input state.
+	var/static/list/client/queue = list()
+
+/datum/controller/subsystem/typing/Recover()
+	status = list()
+	queue = list()
+
+/datum/controller/subsystem/typing/fire(resumed, no_mc_tick)
+	if(!resumed)
+		queue = list()
+		for (var/client/client as anything in global.clients)
+			queue += client
+		if(!length(queue))
+			return
+	var/cut_until = 1
+	var/list/entry
+	for (var/client/client as anything in queue)
+		++cut_until
+		if(QDELETED(client))
+			continue
+		entry = get_client_record(client)
+		if(!islist(entry) || !entry[INDEX_PREFERENCE])
+			continue
+		if(!entry[INDEX_INFLIGHT])
+			update_indicator_state_from_winget(client, entry)
+		else if(world.time < entry[INDEX_TIMEOUT])
+			entry[INDEX_INFLIGHT] = FALSE
+		if(no_mc_tick)
+			CHECK_TICK
+		else if(MC_TICK_CHECK)
+			queue.Cut(1, cut_until)
```

Comment:
Feel like this is a needlessly complicated queue design. Why not just while .len/pop like dozens of other SS, given that you clear the whole queue each time anyway?

Reply:
Well, if you like it you can do it like this, or ask Lohikar for opinions. My go-to more performant variant is just an index you iterate using, no list ops at all. But that probably would have an improvement on the same order as e.g. removing the client data getter proc: not much.